src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/X500NameEncoder.cs

```diff
@@ -0,0 +1,192 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Security.Cryptography;
+using System.Security.Cryptography.X509Certificates;
+using System.Text;
+
+using Microsoft.Win32.SafeHandles;
+
+namespace Internal.Cryptography.Pal
+{
+    internal static class X500NameEncoder
+    {
+        private static readonly char[] s_quoteNeedingChars =
+       {
+            ',',
+            '+',
+            '=',
+            '\"',
+            '\n',
+            // \r is NOT in this list.
+            '<',
+            '>',
+            '#',
+            ';',
+        };
+
+        internal static string X500DistinguishedNameDecode(
+            byte[] encodedDistinguishedName,
+            bool printOid,
+            X500DistinguishedNameFlags flags)
+        {
+            bool reverse = flags.HasFlag(X500DistinguishedNameFlags.Reversed);
+            bool quoteIfNeeded = !flags.HasFlag(X500DistinguishedNameFlags.DoNotUseQuotes);
+            string dnSeparator;
+
+            if (flags.HasFlag(X500DistinguishedNameFlags.UseSemicolons))
```

Comment:
Last I checked, HasFlag was unfortunately fairly expensive.  It'd be better to do the check yourself with something like:
```C#
if ((flags & X500DistinguishedNameFlags.UseSemicolons) == X500DistinguishedNameFlags.UseSemicolons)
```